File: src/main/java/leetcode/Problem1256.java
package leetcode;

/**
 * https://leetcode.com/problems/encode-number/
 */
public class Problem1256 {
    public String encode(int num) {
        String answer = "";
        int n = 1;
        int i = 1;
        outer:
        while (true) {
            for (int j = 0; j < Math.pow(2, n) && i <= num; j++) {
                if (i == num) {
                    answer = Integer.toBinaryString(j);
                    int length = answer.length();
                    for (int k = 0; k < n - length; k++) {
                        answer = "0" + answer;
                    }
                    break outer;
                }
                i++;
            }
            n++;
        }
        return answer;
    }

    public static void main(String[] args) {
        Problem1256 prob = new Problem1256();
        System.out.println(prob.encode(4)); // "01"
        System.out.println(prob.encode(8)); // "001"
        System.out.println(prob.encode(23)); // "1000"
        System.out.println(prob.encode(107)); // "101100"
        System.out.println(prob.encode(1000000000)); // "11011100110101100101000000001"
    }
}

File: src/main/rust/src/bin/problem1256.rs
// https://leetcode.com/problems/encode-number/
pub fn encode(num: i32) -> String {
    if num == 0 {
        return "".to_string();
    }
    let mut n = 1;
    let mut i: u32 = 1;
    while n <= num as u64 {
        n += 2_u64.pow(i);
        i += 1;
    }
    let n = n - 2_u64.pow(i - 1);
    let diff = format!("{:b}", num as u64 - n);
    let mut answer = String::new();
    if diff.len() < i as usize - 1 {
        for i in 0..((i as usize - 1) - diff.len()) {
            answer.push('0');
        }
    }
    answer.push_str(&diff);
    answer
}

fn main() {
    println!("{}", encode(0)); // ""
    println!("{}", encode(1)); // "0"
    println!("{}", encode(2)); // "1"
    println!("{}", encode(3)); // "00"
    println!("{}", encode(4)); // "01"
    println!("{}", encode(5)); // "11"
    println!("{}", encode(6)); // "111"
    println!("{}", encode(7)); // "000"
    println!("{}", encode(8)); // "001"
    println!("{}", encode(23)); // "1000"
    println!("{}", encode(107)); // "101100"
    println!("{}", encode(1000000000)); // "11011100110101100101000000001"
    println!("{}", encode(33195842)); // "111110101000011101000011"
}
